### PR TITLE
Async DNS with SRV records

### PR DESCRIPTION
**Pull request by chbatey**
Still need a way to run the test in CI but ready for review.

Includes commit from https://github.com/akka/akka/pull/25203

### CODE REVIEW

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsQuestion.scala
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/internal/DnsQuestion.scala]: `private[akka]`
several places

File: akka-actor/src/main/scala/akka/io/dns/internal/package.scala
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/internal/package.scala]: can the INTERNAL stuff be added here also?

**Review by patriknw — COMMENTED** [review]
looking good, mostly nitpicks (much probably on old code)

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala]: What happens if the UPD messages are dropped and we don't get any reply? Do we retry or is that rare enough that it's enough to timeout?

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (ktoso) [inline akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: AFAIR this is undesirable and they should be tried in order btw, see ticket on akka-dns

File: akka-actor/src/main/scala/akka/io/dns/DnsRecordType.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/DnsRecordType.scala]: Can't hurt

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala]: Perhaps we should retry on a different name server?

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: Yep, it is any 16 bits that is included in the request that'll be copied to the response

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala]: If we get a CommanFailed for a send we can get the id out of the message and remove it.

Also i'll add it to the timeout of the resolve to remove it from the map for the case that the dns server doesn't respond

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala]: After doing some research i've changed this to try the name servers one at a time. If a timeout occurs it moves to the next

**Comment by chbatey** [comment]
Cleaned up commit history,  ready for re-review

File: akka-actor/src/main/scala/akka/io/dns/DnsProtocol.scala
Comment (johanandren) [inline akka-actor/src/main/scala/akka/io/dns/DnsProtocol.scala]: Wouldn't we normally rather have additional companion apply-methods for these message factories? Does it really help to have the class internal and the factories public? Also, the package private type anyways escapes it's privacy in the return type here.

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (johanandren) [inline akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: Huh? I searched the current codebase for `canonicalNames`/`canonicalName` but get no hits for it?

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (johanandren) [inline akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: Concise!

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala
Comment (johanandren) [inline akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala]: If response would be corrupted/bad and parsing fails and throws, does that stop the client?

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala
Comment (johanandren) [inline akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala]: just if-else the second parameter since that is what is different?

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala
Comment (johanandren) [inline akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala]: If request message is lost will we be leaking memory here since there is no response and maybe no command failed?

**Review by johanandren — COMMENTED** [review]
Looks good, some questions and some unreasonable nitpicks added.

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: I do not take credit for this ;)

**Review by ktoso — COMMENTED** [review]
Submitting my WIP review before flying... incomplete though :-/

File: akka-actor/src/main/scala/akka/io/SimpleDnsManager.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/SimpleDnsManager.scala]: Called it inet dns enabled to match up with the log

File: akka-actor/src/main/scala/akka/io/dns/DnsProtocol.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/DnsProtocol.scala]: Yes i think so

Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/DnsProtocol.scala]: Agreed

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: https://github.com/ilya-epifanov/akka-dns/blob/adbaf294db6c08dadc6eef061860e2329a43b62a/src/main/scala/akka/io/AsyncDnsResolver.scala#L112

File: akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/internal/DnsClient.scala]: `AsyncDnsResolver` times out and sends a DropRequest.

File: akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala
Comment (chbatey) [reply akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala]: We could use a test stub for the DnsClient and then only the DnsClient isn't tested.

File: akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala
Comment (chbatey) [reply akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala]: Should be pending due to the above TODO. Will add in the stubbed DnsClient

File: akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala
Comment (chbatey) [reply akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala]: Doh

File: akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala
Comment (chbatey) [reply akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala]: :+1:

File: akka-actor/src/main/scala/akka/io/Dns.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/Dns.scala]: Wanted an ExtendedActorSystem available to the impl so it can use the dynamcAccess to do the getNameServers reflection

File: akka-actor/src/main/scala/akka/io/dns/DnsProtocol.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/DnsProtocol.scala]: I can't think of a good way, name wise, for this to co-exist with the old one

File: akka-actor/src/main/scala/akka/io/dns/DnsRecordClass.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/DnsRecordClass.scala]: Yep, it is in a dns package

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (chbatey) [reply akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: Or we could remove the feature that we do something with IPs passed to DNS?

File: akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala
Comment (ktoso) [reply akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala]: 👍

File: akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala
Comment (ktoso) [inline akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala]: need to address this or not really? keeping the todo or not?

**Review by johanandren — APPROVED** [review]
LGTM, but has a mima file conflict now

File: akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala
Comment (chbatey) [reply akka-actor-tests/src/test/scala/akka/io/dns/AsyncDnsResolverSpec.scala]: I am going to create a stubbed test for now before I merge this

File: akka-actor/src/main/scala/akka/io/Dns.scala
Comment (ktoso) [reply akka-actor/src/main/scala/akka/io/Dns.scala]: okey

**Comment by chbatey** [comment]
Pushed some stubbed tests. Be good to get a final review @ktoso

File: akka-actor/src/main/scala/akka/io/dns/DnsSettings.scala
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/DnsSettings.scala]: use `DnsFallbackPort`

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: `NonFatal` ?

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: NonFatal

File: akka-actor/src/main/scala/akka/io/dns/internal/DomainName.scala
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/internal/DomainName.scala]: internal api markers

File: akka-actor/src/main/scala/akka/io/dns/internal/RecordClassSerializer.scala
Comment (patriknw) [inline akka-actor/src/main/scala/akka/io/dns/internal/RecordClassSerializer.scala]: internal api markers, check everything in the internal package

**Review by patriknw — APPROVED** [review]
LGTM, can be merged after fixing some minor things

**Comment by johanandren** [comment]
Fixes #12591

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (huntc) [reply akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: Lots of great work here. Perhaps it’d be nice to recognise any prior art here though.

Comment (ktoso) [reply akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: The license headers do so: https://github.com/akka/akka/pull/25214/files#diff-ff500718711617ce0d811010e3075540R3 Where specifically do you mean otherwise.

**Comment by huntc** [comment]
Thanks Konrad. I didn’t notice that.

File: akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala
Comment (huntc) [reply akka-actor/src/main/scala/akka/io/dns/internal/AsyncDnsResolver.scala]: Thanks Konrad. I didn’t notice that.

**Pull request merged**
